### ISSUE TITLE
How to handle patches 

### ISSUE THREAD

**Issue by ChayaSt** [pandegroup/openmm#1335]
@jchodera, @swails and I are working on converting Charmm forcefield files to ffxml using ParmEd here:
https://github.com/pandegroup/openmm/issues/1281 I'm opening this issue to discuss how we should handle patches. In the Charmm polar ff NTER and CTER are combined with every residue but that can become cumbersome with all patches that exist in charmm.

**Comment by jchodera** [comment]
Maybe you could give some more details on this?

How many patches x residues would there be, for example?

Are their drawbacks to expanding all combinations besides the number of
possible combinations?

What kinds of straightforward applications is patching used for?

I think we have discussed these in person, but the others may not be
familiar with the issues.

**Comment by swails** [comment]
> In the Charmm polar ff NTER and CTER are combined with every residue but that can become cumbersome with all patches that exist in charmm.

CHARMM has a lot of patches, and does at least one thing with a patch that OpenMM just does automatically I think (adding disulfide bonds). I've often found that the best way to handle something doesn't reveal itself until there's a substantive use-case.  Can you detail more precisely what you're trying to do (with which patches) outside the simple NTER/CTER patches?

**Comment by ChayaSt** [comment]
> What kinds of straightforward applications is patching used for?

- Post translational modifications (phosphates, sugars, lipids)
- Disulfide bonds
- N and C termini, standard and nonstandard
- Heme ligation
- protonation states
- linking for cyclic proteins and sugars

> Are their drawbacks to expanding all combinations besides the number of
> possible combinations?

One of the issues is how to organize the force fields. For example, the lipids are paramterized in the lipid forcefield and the patch is a way to combine the protein and lipid forcefields. 

There are two kinds of patch behavior. 
1. A modification is added to one residue 
2.  Linking two residues 

How can we handle the patches:
- We can have a ```charmm36_terminal_patches.xml``` and ```charmm36_ptms.xml``` files and when these are loaded all different combinations are tried and only the ones that succeed are loaded into the expanded internal residue templates. The question here is how we will handle linking patches. What should the template look like?
- Or when calling create system, the user can specify a dict with patches and which residue/s to apply it to. Only those combinations will be generated. I think this one might be cleaner and faster since less combinations will be tried.

**Comment by jchodera** [comment]
As @ChayaSt points out, the fact that patches can span two residues makes it difficult to use the standard OpenMM approach of pre-combining every possible patch with every possible patchable residue.  This quickly leads to an explosion of different potential patched residues, but more importantly, residues that become linked by patching still span two separate residue numbers meaning that our strategy of pre-enumerating all possible patched residues no longer works since OpenMM matches by whole residues.  

It may be possible to pre-enumerate all possible patched linked residues and cut them at the residue boundary to form templates, but if there ends up being multiple templates that have the same atoms and internal/external connectivity but different partial charges, this strategy won't work.

An alternative @ChayaSt suggests is that we could extend `ffxml` to allow definition of patches and have the user specify which patches should be applied, but this doesn't fit in cleanly to `ForceField.createSystem()` because it also modifies the `Topology`, so we would need some way to retrieve the patched topology.  Alternatively, we could have `Modeller` apply the patches and have some way of passing this information on to `ForceField`.

We're open to other good ideas here!

**Comment by peastman** [comment]
What about allowing a single template to span two consecutive residues?

**Comment by jchodera** [comment]
> What about allowing a single template to span two consecutive residues?

That could work if the number of possible combinations is not insane.

@ChayaSt: How many residue x patch x residue combinations do you think there are?  Order of magnitude estimate?

**Comment by ChayaSt** [comment]
> @ChayaSt: How many residue x patch x residue combinations do you think there are? Order of magnitude estimate?
residue x patch x residue (disulfide bond and cyclic protein (which is ligated with a glycine at either the N or C terminal)): ~ < 100.
Patches that don't span two residues: ~1000.

**Comment by jchodera** [comment]
What about the attachment of sugars, dyes, or other post-translational modifications?

**Comment by jchodera** [comment]
One of the use cases that @ChayaSt has is parameterizing dyes that get covalently attached by patches.
If we go the route of allowing residue matches that span multiple (potentially non-consecutive, but bonded) residues, it may be useful to have `ForceField` internally expand `residue x patch` and `residue x patch x residue` to allow additional molecules and patches to be specified, rather than requiring these combinations be pre-enumerated.

**Comment by ChayaSt** [comment]
In most cases, patches modify the topology by creating bonds or adding / removing protons. 
* as @jchodera suggested, ```Modeller``` can be used to make the necessary changes to the topology. Is this something that should be done before ```ForceField``` is instantiated or after the patches are applied?
*  we have ```ForceField``` internally expand ```residue x patch ``` and  ```residue x patch x residue ```. Pre-enumerating all combinations will result in thousands of residues. We can have the user specify which patches to expand. 

> What about the attachment of sugars, dyes, or other post-translational modifications?

sugars that span more than one residue: ~1,000's
other patches for sugars: ~100s
Other PMT's: ~100s

**Comment by jchodera** [comment]
Any more thoughts on patches?  We won't get *anywhere* with converting anything here without deciding what to do with patches since N- and C-terminal protein ends are handled with patches!

Can we maybe have a quick video chat about how to handle this?

**Comment by jchodera** [comment]
From our brainstorming over here, this is the simplest way we could think of handling these:
```python
# Load CHARMM36 forcefield (including patch definitions)
ff = ForceField('charmm36.xml')
# Create Modeller instance
m = Modeller(topology)
# Now apply the specified patches defined in the forcefield to change the topology.
# Patch definitions are taken from ForceField object.
# Note that we could have `NTER` and `CTER` handled automatically.
patches = [ ('NTER', 1), ('JOINDYE', 57, 351), ('CTER' : 350) ]
m.applyPatches(ff, patches)
# Now create the System, having ForceField add residue definitions only for patches that are specified.
system = ff.createSystem(patches=patches)
```
We're not clear on what the best syntax is for specifying the patches since `Topology` organizes systems by `(chain, residue)`.  Perhaps the user can either specify a residue ID or a `(chain, resid)` tuple?

**Comment by peastman** [comment]
What would the definitions of those patches look like?  What are all the things that a patch can do?

**Comment by jchodera** [comment]
Here is the [CHARMM patch documentation](http://www.charmm.org/documentation/c34b1/struct.html#%20Patch)

Here are some simple patches that only modify one residue:
https://github.com/ChayaSt/Charmm36_FF/blob/master/toppar/top_all36_prot.rtf#L1279-L1616

Here are some patches that modify two residues to add disaccharide linkages:
https://github.com/ChayaSt/Charmm36_FF/blob/5c689fe83f796a98c42802c18bd2bc11fa64a136/toppar/stream/carb/toppar_all36_carb_model.str#L711-L964

**Comment by jchodera** [comment]
We would presumably have an `ffxml` equivalent specification that specifies how many residues the patch applies to, which atoms to delete and which atoms to connect.  We may not need to specify which bond, angle, and torsion terms to add since these could be discovered automagically.

The `IC` cards allow CHARMM to build missing atoms in ideal geometry if they don't exist.  I don't think we'd need those, since we would presume the user has a structure with all positions already present.

**Comment by peastman** [comment]
I'm afraid that documentation is a little less than illuminating. :)

In the example you gave above, you're using the patch for two different purposes: to modify a Topology and to modify a ForceField.  But if I understand correctly, those are largely separate features, and modifying the ForceField is the important one here.  In your example, you used patches to modify the N and C terminal residues.  But usually you'll be starting from a structure that already has properly terminated residues (PDBFixer takes care of that, as do various other programs).  So the important feature we're talking about here is making sure those modified residues will be recognized by the ForceField.

Considering only that use for the moment, the basic job of a patch is to modify residue templates.  It can do all(?) of the following:

- Add atoms
- Delete atoms
- Add bonds
- Delete bonds
- Add external bonds
- Delete external bonds
- Modify existing atoms to change their types?

From earlier discussion, we also need patches that modify multiple bonded residues at once, right?  Will those residues always be consecutive on the same chain?  Or do we need to handle other cases?

**Comment by ChayaSt** [comment]
> From earlier discussion, we also need patches that modify multiple bonded residues at once, right? Will those residues always be consecutive on the same chain? Or do we need to handle other cases?

Those residues won't necessarily be consecutive or on the same chain.

**Comment by peastman** [comment]
> Those residues won't necessarily be consecutive or on the same chain. 

How will the ForceField figure out what sets of residues to apply the patch to?

**Comment by jchodera** [comment]
> I'm afraid that documentation is a little less than illuminating. :)

Certainly can't dispute that!

> In the example you gave above, you're using the patch for two different purposes: to modify a Topology and to modify a ForceField. But if I understand correctly, those are largely separate features, and modifying the ForceField is the important one here. In your example, you used patches to modify the N and C terminal residues. But usually you'll be starting from a structure that already has properly terminated residues (PDBFixer takes care of that, as do various other programs). So the important feature we're talking about here is making sure those modified residues will be recognized by the ForceField.

OK, we can separate the issue of creating the correct `Topology` (which might already be present in a PDB file containing `CONECT` records, though we do need some way to construct this---maybe through PDBFixer) from ensuring modified residues will be recognized by `ForceField`.

Note that the above example actually did two different things:
* **Modified the N- and C-terminal ends.**  The CHARMM forcefield doesn't define distinct terminal residues, so we need to generate new templates for modified forms of existing residues.  CHARMM also uses `PATC`hes to add post-translational modifications (like phosphorylation, methylation, acylation, or protonation state changes.
* **Join a dye molecule to a residue.**  The `'JOINDYE'` patch (which we made up) connects a dye (residue 351) to a protein residue (residue 7) by eliminating some atoms from each residue and bonding them together to make a new entity that spans two non-consecutive residues.

> Considering only that use for the moment, the basic job of a patch is to modify residue templates. 

To be pedantic: "The basic job of a patch is to create additional residue templates that are modifications of one or more existing residue templates".  The difficulty, as @ChayaSt points out, is that these can be non-consecutive.

> How will the ForceField figure out what sets of residues to apply the patch to?

In the example above, we introduced a syntax for the `patches` argument:
```
patches = [ (patchname, resid1[, resid2, [resid3, ...]]), [(patchname, resid1[, resid2, [resid3, ...]]), ...] ]
```
where `patches` consists of a list of one or more tuples of `patchname` followed by one or more residue IDs that refer to the residue(s) the patch should be applied to.  There could certainly be a better way to provide this information, but if the user supplies it, `createSystem()` knows which residue templates it needs to generate and which residues they apply to.

It would certainly be nicer if `createSystem()` could figure this out automatically, of course, but it might be very complicated to do so.  We could in principle come up with a clever way of applying all loaded patches to every compatible residue and caching these combinations as template signatures, but we would still have to accommodate patches that may span non-consecutive residues.  

Perhaps a viable strategy could be to match all templates and then see if the *unmatched* residues could be accommodated by some combination of patches applied to these residues?  The number of unmatched residues is presumably very few.

> It can do all(?) of the following:
>
> Add atoms
> Delete atoms
> Add bonds
> Delete bonds
> Add external bonds
> Delete external bonds

@ChayaSt and I discussed the possibility that a patch applied to two residues could be thought of as altering these residues and then creating two new one-residue templates that could individually be matched through the normal `ForceField` template assignment procedure.  This *might* work, but we were concerned that there might be some cases in which there might be several new single-residue templates with identical elements and bond graphs but different charges or types.  We don't yet have concrete cases where we would run into this.  CHARMM also segregates charges into neutral charge groups and tries not to break up these charge groups, so it may that such a strategy could actually work.

> From earlier discussion, we also need patches that modify multiple bonded residues at once, right? Will those residues always be consecutive on the same chain? Or do we need to handle other cases?

They will not necessarily be consecutive.

**Comment by peastman** [comment]
> Perhaps a viable strategy could be to match all templates and then see if the unmatched residues could be accommodated by some combination of patches applied to these residues?

If we can make that work, it would be a good strategy.  So let's consider it and see what it would involve.

As an example of a complicated situation, there are various dyes that get attached to CYS residues.  I think(?) that for at least some of these, the atoms and bonds within the CYS residue are identical to those in a residue forming a disulfide bond.  And you might well have both of those in a single molecule.  So there would be two different versions with identical atoms and bonds, but different parameters, distinguished based on what other residues they were bonded to.

In the current force fields, we have a template (CYX) for a CYS residue in disulfide bond.  But in this new system, I assume disulfides would be handled with a patch instead?  So in the initial pass, no match would be found for the residues forming disulfides, the ones bound to dye, or the dye itself.  Next we would look through all those residues individually and try to find a single residue patch that would make them fit.  Still no match.  Then we would identify ones that were bound to each other and look for 2-residue patches that could be applied.  This time we would find matches, and so the atom types would be set.

Is there ever a situation where a 2-residue patch would need to take precedence over a 1-residue patch, or even an ordinary template?  For example, a case where something gets attached to the end of a chain at the usual place, just like an amino acid, but it alters atom types in the residue it's bound to?

**Comment by peastman** [comment]
Any thoughts on my question?

> Is there ever a situation where a 2-residue patch would need to take precedence over a 1-residue patch, or even an ordinary template? For example, a case where something gets attached to the end of a chain at the usual place, just like an amino acid, but it alters atom types in the residue it's bound to?

**Comment by ChayaSt** [comment]
> Is there ever a situation where a 2-residue patch would need to take precedence over a 1-residue patch, or even an ordinary template?

You can have a situation where a dye (or some other PTM) gets attached to the first or last residue. In that case you will need both NTER/CTER patch and the 2 residue patch. It might also alter the atom type in the residue its bound to (depending on what kind of bond is formed).

**Comment by peastman** [comment]
So we might actually need to apply multiple patches to a single residue?  Ok, that's going to be... interesting. :)

I think the big question we need to answer is, "How does it decide what patches to apply to what residues?"  Any suggestions?

**Comment by swails** [comment]
Well the terminal patches are applied for chain-terminating residues as defined in the residue template itself.  So that one is handled "automatically".  The post-translational modification patches are ones I have no experience with, so I'm not sure.

**Comment by ChayaSt** [comment]
This is what some patches look like in the CHARMM files
This is the disulfide bond patch 
https://github.com/ChayaSt/Charmm36_FF/blob/master/toppar/top_all36_prot.rtf#L1693-L1707
The atom type and partial charge on the sulfur changes from atom type S and partial charge -0.23 to atom type SM and partial charge -0.08. In many patches this is what happens; both atom types and partial charges can change. 

Another issue is how to decide which patches to list for each residue and vice versa.  The documentation in the CHARMM files are not consistent. Sometimes it lists which residues you can apply the patches to, example here (GLY and PRO N-terminal):
https://github.com/ChayaSt/Charmm36_FF/blob/master/toppar/top_all36_prot.rtf#L1296-L1328

Other times it lists where not to use the patch (do not use to create dipeptides)
https://github.com/ChayaSt/Charmm36_FF/blob/master/toppar/top_all36_prot.rtf#L1330-L1339

Sometimes it writes the full residue name in the comment section and sometimes the 3 letter code. 

In the CGenFF files, it sometimes lists the residues as core residue:
https://gist.github.com/ChayaSt/ab368040b1ab28991b6e#file-gistfile1-txt-L1-L7
or as to be patched with residue ___:
https://gist.github.com/ChayaSt/ab368040b1ab28991b6e#file-gistfile1-txt-L4741-L4751
Sometimes it doesn't list which residue the patch can be applied to:
https://gist.github.com/ChayaSt/ab368040b1ab28991b6e#file-gistfile1-txt-L8211
While it deletes a H31 (which means that this patch will probably work with any molecule that has an atom names H31), from the graph it seems like it should only attach to Pyrrolidine. 

I didn't look through all the files for all different cases, but it gets complicated in the carb force field. See this file:
https://github.com/ChayaSt/Charmm36_FF/blob/master/toppar/top_all36_carb.rtf and search for PRES
@swails, any ideas?

**Comment by peastman** [comment]
Here's another strange case.  I came across this when converting the CHARMM polarizable force field.  At https://github.com/ChayaSt/Charmm36_FF/blob/master/toppar/drude/drude_toppar_2015_7/toppar_drude_master_protein_2013c.str#L240-L242 it specifies

```
!DEFA FIRS NTER LAST CTER 
!reset default patches 
DEFA FIRS NONE LAST NONE 
```

That means that by default it should not apply any patches at the ends of the chains.  It is explicitly turning off the default NTER and CTER patches.  In spite of that, this force field nonetheless uses NTER and CTER patches at the ends of chains, though that is not specified anywhere in the parameter file!  I never got a coherent explanation of this.

**Comment by jchodera** [comment]
@peastman: Any progress patch handling?  The CHARMM forcefield conversion is dead in the water until this is resolved.  @ChayaSt has put a huge amount of effort into this conversion, and it would be useful if that could be resolved soon via a route to handle patches.

**Comment by peastman** [comment]
What were you expecting me to do?  I was the last person who posted here, and there still seem to be a lot of design issues that haven't been resolved.

**Comment by jchodera** [comment]
My recollection was that we had a video chat in which we collectively decided you would explore some options for how CHARMM-style patches could best be handled within the framework of `ForceField.`  If you still need clarification for how some of these things work in CHARMM, @ChayaSt can field those questions.

**Comment by peastman** [comment]
Please see the above discussion.  We've had a lot of additional discussion since the video chat, and much of it is about questions that are still open regarding the requirements.

**Comment by jchodera** [comment]
@ChayaSt : Can you chime in here and answer the open questions about requirements?

**Comment by ChayaSt** [comment]
Most of the open questions are how to deal with inconsistencies in the CharMM files. We want to be able to list all the patches a residue can have and vice versa, but it's not clear form CharMM documentation how to handle that because they don't have consistent documentation. @peastman, I linked most cases here https://github.com/pandegroup/openmm/issues/1335#issuecomment-179529573.

To start, let's figure out how to deal with the cases in the above comment. It's not exhaustive but encompasses most cases. I will go through all the files and link different cases here.

**Comment by peastman** [comment]
It might be useful to start imagining what the definitions might look like.  For example, here's how we might define the N-terminal residue patch:

```XML
<ResiduePatch name="NTER">
  <RemoveAtom name="HN"/>
  <ChangeAtom name="N" type="NH3" charge="-0.3"/>
  <ChangeAtom name="CA" type="CT1" charge="0.21"/>
  <ChangeAtom name="HA" type="HB1" charge="0.1"/>
  <AddAtom name="H1" type="HC" charge="0.33"/>
  <AddAtom name="H2" type="HC" charge="0.33"/>
  <AddAtom name="H3" type="HC" charge="0.33"/>
  <AddBond atomName1="N" atomName2="H1"/>
  <AddBond atomName1="N" atomName2="H2"/>
  <AddBond atomName1="N" atomName2="H3"/>
  <RemoveExternalBond atomName="N"/>
</ResiduePatch>
```

Hopefully that's pretty self explanatory.  Now consider the patch for disulfides.  That's more complicated because it involves two residues.  It might look something like this:

```XML
<ResiduePatch name="DISU" residues="2">
  <RemoveAtom name="1:HG1"/>
  <RemoveAtom name="2:HG1"/>
  <ChangeAtom name="1:CB" type="CT2" charge="-0.1"/>
  <ChangeAtom name="2:CB" type="CT2" charge="-0.1"/>
  <ChangeAtom name="1:SG" type="SM" charge="-0.08"/>
  <ChangeAtom name="2:SG" type="SM" charge="-0.08"/>
  <AddBond atomName1="1:SG" atomName2="2:SG"/>
</ResiduePatch>
```

I've told it this patch involves two residues with `residues="2"`, and every atom name is prefixed by `1:` or `2:` to tell it which of the two residues it relates to.

Now we need to tell it which patches can be applied to which residues.  That might look something like this:

```XML
<Residue name="ALA">
  <AllowPatch name="NTER"/>
  <AllowPatch name="CTER"/>
  ...
</Residue>
```

For the disulfide patch it's a little more complicated.  That involves two residues.  In this case they happen to be identical and are treated symmetrically, but that need not always be the case.  So we need to tell it that DISU can be applied to CYS, and also that it should consider the CYS for both of the two residues in the patch.  Perhaps something like this:

```XML
<AllowPatch name="DISU:1"/>
<AllowPatch name="DISU:2"/>
```

**Comment by ChayaSt** [comment]
These sounds good.

We also need to consider cases where a patch can be applied to more than 2 residues. Here's an example of a patch to apply to 3 residues:
https://github.com/ChayaSt/Charmm36_FF/blob/master/toppar/top_all36_carb.rtf#L5304-L5326

We can extend the way you wrote the DISU patch so that is should look something like this:

```
<ResiduePatch name="MELZ" residues=3>
  <RemoveAtom name="1:HO1"/>
  <RemoveAtom name="2:O2"/>
  <RemoveAtom name="2:HO2/">
  <RemoveAtom name="3:HO1/">
  <RemoveAtom name="2:O3/">
  <RemoveAtom name="2:HO3/">
  <ChangeAtom name="1:C1" type="CC3162" charge="0.29"/>
  <ChangeAtom name="1O1" type="OC302" charge="-0.36"/>
  <ChangeAtom name="2:C2" type="CC3051" charge="0.38"/>
  <ChangeAtom name="3:C1" type="CC3162" charge="0.29"/>
  <ChangeAtom name="3:O1" type="OC303" charge="-0.36"/>
  <ChangeAtom name="2:C3" type="CC3151" charge="0.09"/>
  <AddBond atomName1="1:O1" atomName2="2:C2"/>
  <AddBond atomName1="3:O1" atomName2="C3/">
</ResiduePatch>
```
This patch is applied to RESI  AGLC, BFRU, AGLC. So we can do something like this for RESI AGLC:

```
<AllowPatch name="MELZ:1"/>
<AllowPatch name="MELZ:2"/>
<AllowPatch name="MELZ:3"/>
```

and this for BFRU
```
<AllowPatch name="MELZ:1"/>
<AllowPatch name="MELZ:3"/>
```
It gets even more complicated when the patch is applied to 3 different residues:
https://github.com/ChayaSt/Charmm36_FF/blob/master/toppar/top_all36_carb.rtf#L5328-L5355

**Comment by peastman** [comment]
Or rather, AGLC would specify

```XML
<AllowPatch name="MELZ:1"/>
<AllowPatch name="MELZ:3"/>
```

And BFRU would specify

```XML
<AllowPatch name="MELZ:2"/>
```

**Comment by ChayaSt** [comment]
Yes, that makes more sense. 

With the 3 different residues, I assume that the order is as documented AGLC,BFRU,AGAL. So we can do something like this:

For AGLC

```
<AllowPatch name="RAFF:1"/>
```
for BFRU
```
<AllowPatch name="RAFF:2"/>
```
and AGAL
```
<AllowPatch name="RAFF:3"/>
```

**Comment by peastman** [comment]
I'd like to start working on this.  Is everyone ok with the design described above?

The hard part will be automatically figuring out what patches to apply.  I'm still not 100% convinced that can always be done in an unambiguous way.  Perhaps we could start putting together a list of test cases, with the most complicated, convoluted combinations of patches we ever expect to have to support?

**Comment by ChayaSt** [comment]
> Perhaps we could start putting together a list of test cases, with the most complicated, convoluted combinations of patches we ever expect to have to support?

I can start working on this.

**Comment by peastman** [comment]
Thanks!  That will be great.  Things like attaching a dye to a terminal residue that also forms a disulfide bond... as strange as you can make it. :)

**Comment by peastman** [comment]
I have it working with single-residue patches.  Now I'm trying to implement the matching for multi-residue patches, and I'm realizing that if I do it in a completely general way, it can produce a completely insane combinatorial explosion.  It needs to consider all possible combinations of patches, each applied to all possible combinations of templates, each of those matched against all possible sets of residues.  I'm trying to find a way to cut back on that.

Most of the time it shouldn't actually be a problem.  There will only be a small number of residues that don't match one of the basic templates, and most of the remaining ones will only require a single one-residue patch.  But if something goes wrong, it could go very wrong.

For example, suppose you have a protein force field with templates for all the standard amino acids, and a bunch of 1-, 2-, and 3- residue patches to describe terminal residues, disulfide bonds, various kinds of post-translational modifications, etc.  And now you accidentally give it a topology containing a DNA chain.  It won't be able to match it, of course, since the force field doesn't describe DNA.  But to figure that out, it has to try all possible combinations of patches applied to all possible combinations of templates matched against all possible subsets of residues.  So your script just hangs while it starts trying to churn through 10^50 different combinations.  That's the sort of thing I'm afraid of.

Would it be reasonable to assume that no residue will ever be affected by more than one multi-residue template?  It could still have any number of one-residue templates.  But it doesn't need to consider the possibility of multiple overlapping multi-residue patches, where one patch gets applied to residues 1 and 2, and another one gets applied to residues 2 and 3?

I know my previous message gave an example of exactly this sort of situation: a dye attached to a residue that also forms a disulfide bond.  Is that actually a realistic situation?  Do we need to be able to deal with that case?

**Comment by ChayaSt** [comment]
> I know my previous message gave an example of exactly this sort of situation: a dye attached to a residue that also forms a disulfide bond. Is that actually a realistic situation? Do we need to be able to deal with that case?

I was going to point this out, but this is not a realistic situation. The dye is attached to the Sulfur so it can't form disulfide bonds. 

> Would it be reasonable to assume that no residue will ever be affected by more than one multi-residue template? It could still have any number of one-residue templates. But it doesn't need to consider the possibility of multiple overlapping multi-residue patches, where one patch gets applied to residues 1 and 2, and another one gets applied to residues 2 and 3?

From the patches I have looked out, I didn't see such a case. But it might happen with the carb forcefield because there you're forming these chains. What if you also want to attach it to a protein? So you'll have the one carb residue that has a multi-residue patch to attach to the protein and another multi-residue patch to attach another monosaccharide.

**Comment by peastman** [comment]
Is it actually necessary to use patches to form carbohydrate chains?  Could that be done with ordinary templates and external bonds, the way it is for proteins and nucleotides?

**Comment by ChayaSt** [comment]
> Is it actually necessary to use patches to form carbohydrate chains? Could that be done with ordinary templates and external bonds, the way it is for proteins and nucleotides?

This is how the the carb force field is structured as you can see [here](https://github.com/ChayaSt/Charmm36_FF/blob/master/toppar/top_all36_carb.rtf)

The patches are used to build the di and polysaccharides. Maybe they structured it this way because there can be more than one position where the residues can be linked unlike proteins and nucleotides? Yes, it can get crazy indeed...

**Comment by peastman** [comment]
Does that mean you could get long chains of overlapping patches?  One patch applied to residues 1 and 2, another to 2 and 3, another to 3 and 4, and so on?  If so, I'm really not sure we can support that.  It pretty much requires the combinatorial explosion of possibilities I mentioned before.

**Comment by peastman** [comment]
Any progress on this?  I'd really like to merge #1453, but we really need more tests first.

**Comment by peastman** [comment]
This has been implemented.
